### PR TITLE
Issue 40659: Security fixes for vulnerabilities found in security scan report

### PR DESCRIPTION
**Pull request by labkey-ankurj**
#### Rationale
This PR addresses the security concerns mentioned in this issue https://www.labkey.org/home/Developer/issues/Secure/issues-details.view?issueId=40659

#### Related Pull Requests
* https://github.com/LabKey/targetedms/pull/209

### CODE REVIEW

File: core/src/org/labkey/core/login/LoginController.java [hunk 1]
```diff
@@ -673,6 +673,14 @@ public Object execute(LoginForm form, BindException errors)
                 response = new ApiSimpleResponse();
                 response.put("success", true);
 
+                Cookie emailCookie = PageFlowUtil.getCookie(request, "email");
+
```
Comment (labkey-jeckels) [inline core/src/org/labkey/core/login/LoginController.java]: As discussed, let's handle this in the JS code that is the only spot that deals with the cookie right now.

server/modules/platform/core/webapp/login.js

and only set it to be secure if the page is an "https://" URL

**Review by labkey-jeckels — CHANGES_REQUESTED** [review]

File: core/src/org/labkey/core/login/LoginController.java (hunk 1 above)
Comment (labkey-ankurj) [reply core/src/org/labkey/core/login/LoginController.java]: Done

File: api/webapp/clientapi/core/Utils.js [hunk 2]
```diff
@@ -388,6 +388,10 @@ LABKEY.Utils = new function()
             if (pageonly)
                 path = location.pathname.substring(0, location.pathname.lastIndexOf('/'));
             document.cookie = name + "=" + value + expires + "; path=" + path;
+
+            if (name === "email" && window.location.href.startsWith("https")) {
```
Comment (labkey-jeckels) [inline api/webapp/clientapi/core/Utils.js]: To be extra cautious, do a .toLowerCase() here

File: core/resources/views/_footer.html [hunk 3]
```diff
@@ -1,3 +1,3 @@
 <div style="text-align: center; padding:6px;">
-    <a href="http://labkey.com" target="_blank">Powered by LabKey</a>
+    <a href="http://labkey.com" target="_blank" rel="noopener">Powered by LabKey</a>
```
Comment (labkey-jeckels) [inline core/resources/views/_footer.html]: Let's make this https instead of http

Comment (labkey-jeckels) [reply core/resources/views/_footer.html]: Also, looks like this should be "noopener noreferrer". https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/

**Review by labkey-jeckels — CHANGES_REQUESTED** [review]
Please also check for other places in the code where we create links with target="_blank". It should be safe to always add the new attribute, but it's only required when we target some external server (not a link to another page on the same server)  For example, I found one in server/modules/platform/core/src/org/labkey/core/admin/customizeEmail.jsp

We should also make our link rendering inject this in server/modules/platform/api/src/org/labkey/api/data/DataColumn.java, server/modules/platform/api/src/org/labkey/api/data/SimpleDisplayColumn.java, and server/modules/platform/api/src/org/labkey/api/data/DataRegion.java in their renderGridCellContents() methods.

**Comment by labkey-ankurj** [comment]
> Please also check for other places in the code where we create links with target="_blank". It should be safe to always add the new attribute, but it's only required when we target some external server (not a link to another page on the same server) For example, I found one in server/modules/platform/core/src/org/labkey/core/admin/customizeEmail.jsp
> 
> We should also make our link rendering inject this in server/modules/platform/api/src/org/labkey/api/data/DataColumn.java, server/modules/platform/api/src/org/labkey/api/data/SimpleDisplayColumn.java, and server/modules/platform/api/src/org/labkey/api/data/DataRegion.java in their renderGridCellContents() methods.

Aren't the urls in renderGridCellContents of these classes targeting the same server?

**Comment by labkey-jeckels** [comment]
99% of the time, yes, but they can be external URLs too, like http://foo.com/path?id=${RowId}

File: api/webapp/clientapi/core/Utils.js [hunk 4]
```diff
@@ -389,7 +389,7 @@ LABKEY.Utils = new function()
                 path = location.pathname.substring(0, location.pathname.lastIndexOf('/'));
             document.cookie = name + "=" + value + expires + "; path=" + path;
 
-            if (name === "email" && window.location.href.startsWith("https")) {
+            if (name.toLowerCase() === "email" && window.location.href.startsWith("https")) {
```
Comment (labkey-jeckels) [inline api/webapp/clientapi/core/Utils.js]: I was intending for the lower casing to be on window.location.href, not the cookie name

File: api/src/org/labkey/api/data/DataColumn.java [hunk 5]
```diff
@@ -409,8 +409,9 @@ public void renderGridCellContents(RenderContext ctx, Writer out) throws IOExcep
                     out.write("\" style=\"");
                     out.write(css);
                 }
-                
-                out.write("\">");
+
+                out.write("\" rel=\"noopener noreferrer\"");
```
Comment (labkey-jeckels) [inline api/src/org/labkey/api/data/DataColumn.java]: One more tweak - let's only do this when we have a target window

File: api/src/org/labkey/api/data/SimpleDisplayColumn.java [hunk 6]
```diff
@@ -144,7 +144,8 @@ public void renderGridCellContents(RenderContext ctx, Writer out) throws IOExcep
                 out.write(linkCls);
             }
 
-            out.write("'>");
+            out.write("' rel='noopener noreferrer'");
```
Comment (labkey-jeckels) [inline api/src/org/labkey/api/data/SimpleDisplayColumn.java]: Same here -  let's only do this when we have a target window

**Review by labkey-jeckels — APPROVED** [review]
Three more tiny changes, but I'll make it as approved

**Pull request merged**